distributed/shuffle/_rechunk.py

```diff
@@ -0,0 +1,96 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+from dask.base import tokenize
+from dask.highlevelgraph import HighLevelGraph, MaterializedLayer
+
+from distributed.shuffle._shuffle import (
+    ShuffleId,
+    _get_worker_extension,
+    barrier_key,
+    shuffle_barrier,
+)
+
+if TYPE_CHECKING:
+    import numpy as np
+    from typing_extensions import TypeAlias
+
+    import dask.array as da
+
+
+# TODO remove quotes (requires Python >=3.9)
+ChunkedAxis: TypeAlias = "tuple[float, ...]"  # chunks must either be an int or NaN
+ChunkedAxes: TypeAlias = "tuple[ChunkedAxis, ...]"
+NIndex: TypeAlias = "tuple[int, ...]"
+
+
+def rechunk_transfer(
+    input: np.ndarray,
+    id: ShuffleId,
+    input_chunk: NIndex,
+    new: ChunkedAxes,
+    old: ChunkedAxes,
+) -> int:
+    try:
+        return _get_worker_extension().add_partition(
+            input,
+            input_partition=input_chunk,
+            shuffle_id=id,
+            type="ArrayRechunk",
+            new=new,
+            old=old,
+        )
+    except Exception as e:
+        raise RuntimeError("rechunk_transfer failed during shuffle {id}") from e
+
+
+def rechunk_unpack(
+    id: ShuffleId, output_chunk: NIndex, barrier_run_id: int
+) -> np.ndarray:
+    try:
+        return _get_worker_extension().get_output_partition(
+            id, barrier_run_id, output_chunk
+        )
+    except Exception as e:
+        raise RuntimeError("rechunk_unpack failed during shuffle {id}") from e
+
+
+def rechunk_p2p(x: da.Array, chunks: ChunkedAxes) -> da.Array:
+    import numpy as np
+
+    import dask.array as da
+
+    if x.size == 0:
+        # Special case for empty array, as the algorithm below does not behave correctly
+        return da.empty(x.shape, chunks=chunks, dtype=x.dtype)
+
+    dsk: dict = {}
+    token = tokenize(x, chunks)
+    _barrier_key = barrier_key(ShuffleId(token))
+    name = f"rechunk-transfer-{token}"
+    transfer_keys = []
+    for index in np.ndindex(tuple(len(dim) for dim in x.chunks)):
+        transfer_keys.append((name,) + index)
+        dsk[(name,) + index] = (
+            rechunk_transfer,
+            (x.name,) + index,
+            token,
+            index,
+            chunks,
+            x.chunks,
+        )
+
+    dsk[_barrier_key] = (shuffle_barrier, token, transfer_keys)
+
+    name = f"rechunk-p2p-{token}"
+
+    for index in np.ndindex(tuple(len(dim) for dim in chunks)):
+        dsk[(name,) + index] = (rechunk_unpack, token, index, _barrier_key)
```

Comment:
A possible optimization to think about at a later stage is to introduce multiple disconnected barriers for a rechunk. Part of the reason why the current tasks based approach is faster is because it can benefit from colocation but I suspect it's also because rechunk graphs are often not all-to-all but only many-to-many. This could be expressed by using multiple barriers but would make this code much more complex, of course.
This would allow some downstream tasks to continue even before the full rechunk is done.